packages/blade/src/components/Button/BaseButton/BaseButton.tsx

```diff
@@ -0,0 +1,97 @@
+import type { ReactElement } from 'react';
+import getIn from '../../../utils/getIn';
+import BaseText from '../../Typography/BaseText';
+import { useTheme } from '../../BladeProvider';
+import type { IconComponentType } from '../../Icons';
+import StyledBaseButton from './StyledBaseButton';
+
+export type BaseButtonIcon = undefined | null | IconComponentType;
+
+type BaseButtonCommonProps = {
+  variant?: 'primary' | 'secondary' | 'tertiary' | 'link';
+  intent?: 'positive' | 'negative' | 'notice' | 'info' | 'neutral';
+  contrast?: 'low' | 'high';
+  size?: 'large' | 'medium' | 'small' | 'xsmall';
+  iconPosition?: 'left' | 'right';
+  isDisabled?: boolean;
+  isFullWidth?: boolean;
+  onClick?: () => void;
+  type?: 'button' | 'reset' | 'submit';
+};
+
+type BaseButtonWithoutIconProps = BaseButtonCommonProps & {
+  icon?: undefined;
+  children: string;
+};
+
+type BaseButtonWithIconProps = BaseButtonCommonProps & {
+  icon: IconComponentType;
+  children?: string;
+};
+
+export type BaseButtonProps<T> = T extends {
+  icon?: infer Icon;
+}
+  ? Icon extends Extract<BaseButtonIcon, IconComponentType>
+    ? BaseButtonWithIconProps
+    : Icon extends Extract<BaseButtonIcon, undefined | null>
+    ? BaseButtonWithoutIconProps
+    : T
+  : T;
+
+const BaseButton = <T extends { icon: BaseButtonIcon }>({
+  variant = 'primary',
+  intent,
+  contrast = 'low',
+  size = 'medium',
+  icon,
```

Comment:
```suggestion
  icon: Icon,
```

you can title case it inline only